src/FlashLoanRouter.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8;
+
+import {ICowSettlement} from "./interface/ICowSettlement.sol";
+import {IFlashLoanRouter} from "./interface/IFlashLoanRouter.sol";
+import {IFlashLoanSolverWrapper} from "./interface/IFlashLoanSolverWrapper.sol";
+import {BytesUtil, LoanRequest, LoansWithSettlement} from "./library/LoansWithSettlement.sol";
+import {ICowAuthentication} from "./vendored/ICowAuthentication.sol";
+import {IERC20} from "./vendored/IERC20.sol";
+
+contract FlashLoanRouter is IFlashLoanRouter {
+    using BytesUtil for bytes;
+    using LoansWithSettlement for bytes;
+
+    IFlashLoanSolverWrapper NO_PENDING_BORROWER = IFlashLoanSolverWrapper(address(0));
+
+    ICowSettlement public immutable settlementContract;
+    ICowAuthentication public immutable settlementAuthentication;
+
+    IFlashLoanSolverWrapper private transient pendingBorrower;
+    bytes32 private transient pendingDataHash;
+
+    modifier onlySolver() {
+        // Todo: investigate security implication of self calls.
+        require(settlementAuthentication.isSolver(msg.sender), "Only callable by a solver");
+        _;
+    }
+
+    modifier onlyPendingBorrower() {
+        // Todo: investigate security implication of self calls.
+        require(msg.sender == address(pendingBorrower), "Only callable by borrower");
+        _;
+    }
+
+    constructor(ICowSettlement _settlementContract) {
+        settlementContract = _settlementContract;
+        settlementAuthentication = ICowAuthentication(_settlementContract.authenticator());
+    }
+
+    function flashLoanAndSettle(LoanRequest.Data[] calldata loans, bytes calldata settlement) external onlySolver {
+        bytes memory encodedLoansWithSettlement = LoansWithSettlement.encodeLoansWithSettlement(loans, settlement);
+        borrowNextLoan(encodedLoansWithSettlement);
+    }
+
+    function borrowerCallback(bytes memory encodedLoansWithSettlement) external onlyPendingBorrower {
+        pendingBorrower = NO_PENDING_BORROWER;
+        require(keccak256(encodedLoansWithSettlement) == pendingDataHash, "Bad data from borrower");
+        borrowNextLoan(encodedLoansWithSettlement);
+    }
+
+    function borrowNextLoan(bytes memory encodedLoansWithSettlement) private {
+        if (encodedLoansWithSettlement.loansCount() == 0) {
+            settle(encodedLoansWithSettlement.destroyAndExtractSettlement());
+        } else {
+            LoanRequest.Data memory loan = encodedLoansWithSettlement.popLoanRequest();
+            IFlashLoanSolverWrapper borrower = loan.borrower;
+            bytes32 dataHash = keccak256(encodedLoansWithSettlement);
```

Comment:
I wonder if it would be fine to only hash the settlement calldata and check this before we call settle? If the laon requests are tempered with the settlement should probably already fail shouldn't it? Would this allow us to save significant gas?